### PR TITLE
Add missing fsync calls in the snapshot module

### LINKED ISSUE #4846: Missing sync file after receiving snapshot
## Bug Report

Should sync_all at https://github.com/tikv/tikv/blob/master/src/raftstore/store/snap.rs#L1057 when receiving snapshot.

### PR DESCRIPTION
**Pull request by benpigchu**
Signed-off-by: Ben Pig Chu <benpichu@gmail.com>

## What have you changed? (mandatory)

Fix #4846 by add some fsync calls in the snapshot module. Also add util used to call fsync by path

## What are the type of the changes? (mandatory)

- Bug fix 

## How has this PR been tested? (mandatory)

Unit test

## Does this PR affect documentation (docs) or release note? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

### CODE REVIEW

File: components/tikv_util/src/file.rs [hunk 1]
```diff
@@ -46,6 +46,13 @@ pub fn create_dir_if_not_exist<P: AsRef<Path>>(dir: P) -> io::Result<bool> {
     }
 }
 
+/// Call fsync on file or directory by its path
+pub fn fsync_by_path<P: AsRef<Path>>(path: P) -> io::Result<()> {
```
Comment (zhangjinpeng1987) [inline components/tikv_util/src/file.rs]: Please add a unit test for this fn.

**Comment by BusyJay** [comment]
How about calling sync only when the snap is about to be saved?

File: src/raftstore/store/snap.rs [hunk 2]
```diff
@@ -966,6 +972,7 @@ impl Snapshot for Snap {
             meta_file.sync_all()?;
         }
         fs::rename(&self.meta_file.tmp_path, &self.meta_file.path)?;
+        fsync_by_path(&self.dir_path)?;
```
Comment (BusyJay) [inline src/raftstore/store/snap.rs]: I suggest to sync it at L966 first and then here. IO operation can be reordered.

File: src/raftstore/store/snap.rs [hunk 3]
```diff
@@ -646,6 +650,7 @@ impl Snap {
             // to indicate if the sst file is empty.
             if cf_file.kv_count > 0 {
                 fs::rename(&cf_file.tmp_path, &cf_file.path)?;
+                fsync_by_path(&self.dir_path)?;
```
Comment (BusyJay) [inline src/raftstore/store/snap.rs]: I don't think it's necessary. The files will be removed when restarted anyway.

File: components/tikv_util/src/file.rs [hunk 4]
```diff
@@ -190,4 +197,10 @@ mod tests {
         assert!(!create_dir_if_not_exist(&subdir).unwrap());
         assert!(delete_dir_if_exist(&subdir).unwrap());
     }
+
+    #[test]
+    fn test_fsync_by_path() {
+        let tmp_dir = TempDir::new("").unwrap();
+        fsync_by_path(tmp_dir.path()).unwrap();
```
Comment (zhangjinpeng1987) [inline components/tikv_util/src/file.rs]: Please also fsync a unexisted path, and an error is expected .

File: components/tikv_util/src/file.rs (hunk 1 above)
Comment (zhangjinpeng1987) [reply components/tikv_util/src/file.rs]: s/fsync_by_path/sync_dir

File: src/raftstore/store/snap.rs (hunk 2 above)
Comment (benpigchu) [reply src/raftstore/store/snap.rs]: Done

File: components/tikv_util/src/file.rs (hunk 1 above)
Comment (benpigchu) [reply components/tikv_util/src/file.rs]: However this function can also be used to call fsync on a normal file

Comment (zhangjinpeng1987) [reply components/tikv_util/src/file.rs]: Currently no other where use it, please keep the semantics clear.

File: components/tikv_util/src/file.rs (hunk 4 above)
Comment (benpigchu) [reply components/tikv_util/src/file.rs]: Done

File: components/tikv_util/src/file.rs (hunk 1 above)
Comment (benpigchu) [reply components/tikv_util/src/file.rs]: Done

File: src/raftstore/store/snap.rs [hunk 5]
```diff
@@ -667,6 +672,7 @@ impl Snap {
             f.flush()?;
         }
         fs::rename(&self.meta_file.tmp_path, &self.meta_file.path)?;
+        fsync_by_path(&self.dir_path)?;
```
[the remark this answers is not in the document]
Comment (benpigchu) [reply src/raftstore/store/snap.rs]: Removed

File: src/raftstore/store/snap.rs (hunk 3 above)
Comment (benpigchu) [reply src/raftstore/store/snap.rs]: Removed

**Comment by benpigchu** [comment]
PTAL @zhangjinpeng1987 @BusyJay

**Comment by BusyJay** [comment]
I think sync them in `fn save` function is more understandable and not easy to make mistake.

**Comment by benpigchu** [comment]
I reverted the commit that removes the sync in the save functions. PTAL @zhangjinpeng1987 @BusyJay

File: src/raftstore/store/snap.rs [hunk 6]
```diff
@@ -1054,6 +1060,7 @@ impl Write for Snap {
                 file.write_all(&next_buf[0..left])?;
                 digest.write(&next_buf[0..left]);
                 cf_file.written_size += left as u64;
+                cf_file.file.as_mut().unwrap().sync_all()?;
```
Comment (hicqu) [inline src/raftstore/store/snap.rs]: This line can be removed.

**Comment by benpigchu** [comment]
I misunderstand the previous comment, now I move the `sync_all` call in `impl Write` block to `fn save()`.
 PTAL @zhangjinpeng1987 @BusyJay

**Comment by hicqu** [comment]
lgtm.

File: src/raftstore/store/snap.rs [hunk 7]
```diff
@@ -924,6 +926,7 @@ impl Snapshot for Snap {
             {
                 let mut file = cf_file.file.take().unwrap();
                 file.flush()?;
+                file.sync_all()?;
```
Comment (BusyJay) [inline src/raftstore/store/snap.rs]: If the size is 0 like L920, will there be any file generated?

Comment (benpigchu) [reply src/raftstore/store/snap.rs]: Seems no, see L459

**Pull request merged**